src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.Automata.cs

```diff
@@ -0,0 +1,429 @@
+            }
+
+            return state;
+
+            // Assign the context-independent information for the given state
+            static ContextIndependentState BuildStateInfo(int stateId, bool isInitial, bool isDeadend, bool isNullable, bool canBeNullable)
+            {
+                Debug.Assert(stateId > 0);
+                Debug.Assert(!isNullable || canBeNullable);
+
+                ContextIndependentState info = 0;
+
+                if (isInitial)
+                {
+                    info |= ContextIndependentState.IsInitial;
+                }
+
+                if (isDeadend)
+                {
+                    info |= ContextIndependentState.IsDeadend;
+                }
+
+                if (canBeNullable)
+                {
+                    info |= ContextIndependentState.CanBeNullable;
+                    if (isNullable)
+                    {
+                        info |= ContextIndependentState.IsNullable;
+                    }
+                }
+
+                return info;
+            }
+        }
+
+        /// <summary>
+        /// Make an NFA state for the given node and previous character kind.
+        /// </summary>
+        /// <returns>the NFA ID of the new state, or null if the state is a dead end</returns>
+        private int? CreateNfaState(SymbolicRegexNode<TSet> node, uint prevCharKind)
+        {
+            Debug.Assert(Monitor.IsEntered(this));
+            Debug.Assert(node.Kind != SymbolicRegexNodeKind.Alternate);
+
+            // First make the underlying core state
+            DfaMatchingState<TSet> coreState = GetOrCreateState(node, prevCharKind);
+
+            // If the state is a dead end then don't create an NFA state, as dead ends in NFA mode are represented
+            // as empty lists of states.
+            if (coreState.IsDeadend(Solver))
+            {
+                return null;
+            }
+
+            if (!_nfaStateArrayInverse.TryGetValue(coreState.Id, out int nfaStateId))
+            {
+                nfaStateId = MakeNewNfaState(coreState.Id);
+            }
+
+            return nfaStateId;
+        }
+
+        /// <summary>Creates a new NFA state for the underlying core state</summary>
+        private int MakeNewNfaState(int coreStateId)
+        {
+            Debug.Assert(Monitor.IsEntered(this));
+
+            if (NfaStateCount == _nfaStateArray.Length)
+            {
+                // The growth factor 2 matches that of List<T>
+                int newsize = Math.Max(_nfaStateArray.Length * 2, InitialNfaStateCapacity);
+                ConcurrentArrayResize(ref _nfaStateArray, newsize);
+                ConcurrentArrayResize(ref _nfaDelta, newsize << _mintermsLog);
+                ConcurrentArrayResize(ref _capturingNfaDelta, newsize << _mintermsLog);
+            }
+
+            int nfaStateId = NfaStateCount;
+            _nfaStateArray[nfaStateId] = coreStateId;
```

Comment:
The code here is confusing, as `NfaStateCount` is actually `_nfaStateArrayInverse.Count`, so that second assignment actually has relevant "side effects". I'll rewrite this to be clearer.